tremor-pipeline/src/op/trickle/select.rs

```diff
@@ -603,2059 +214,261 @@ impl Operator for TrickleSelect {
-                                    windows.get_mut(emit_window_events.len())
-                                {
-                                    consts.window = Value::from(non_emit_window.name.to_string());
+                        let mut ctx = SelectCtx {
+                            select,
+                            local_stack: &local_stack,
+                            node_meta,
+                            opts,
+                            ctx: &ctx,
+                            event_id: outgoing_event_id,
+                            event_id_gen,
+                            ingest_ns,
+                            op_meta: &op_meta,
+                            origin_uri: &None,
+                            transactional: w.transactional,
+                            recursion_limit,
+                        };
+                        if w.holds_data {
+                            if let Some(port_and_event) =
+                                super::select::execute_select_and_having(&ctx, &env, &data)?
+                            {
+                                res.events.push(port_and_event);
+                            };
+                        }
+                        // re-initialize aggr state for new window
+                        // reset transactional state for outgoing events
 
-                                    // get current window group
-                                    let this_group = stry!(get_or_create_group(
-                                        non_emit_window,
-                                        event_id_gen,
-                                        aggregates,
-                                        group_str,
-                                        group_value,
-                                    ));
-                                    for (this, prev) in
-                                        this_group.aggrs.iter_mut().zip(scratch1.aggregates.iter())
-                                    {
-                                        this.invocable.merge(&prev.invocable).map_err(|e| {
-                                            let r: Option<&Registry> = None;
-                                            e.into_err(prev, prev, r, &node_meta)
-                                        })?;
-                                    }
-                                    this_group.track_transactional(scratch1.transactional);
-                                    this_group.id.track(&event_id_scratch1);
-                                }
-                            }
-                        } else {
-                            // if this happens, we fucked up internally.
-                            return Err("Internal Error. Cannot execute select with multiple windows, no scratches available. This is bad!".into());
+                        if let Some(next) = &mut w.next {
+                            can_remove = next.on_event(
```

Comment:
why calling `on_event` for `next` when it has not been called on `w`? Shouldnt we handle the signal case differently?

Reply:
`w.on_event` would try to accumulate the data, which we don't have, on a select, that would mark the window as tracking data, adding empty values, emitting it, and keeping it around. for signal handling we needed a different logic for `on_event` that circumvent the ingest an event part.

Reply:
Just wanted to check that this is on purpose and not a blind-spot. Happens with huge workloads like this PR (at least to me).

Reply:
It easily does :D yes, and sorry for the size :)